ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -2,305 +2,354 @@
+        excessive_draw = temp_draw;
+      }
 
-  m_power_values[++m_power_values_index % m_power_values.size()] = power_watts;   // [W]
-  auto mean_mechanical_power =
-      accumulate(begin(m_power_values), end(m_power_values), 0.0) / m_power_values.size();
+      m_power_models[i].model.GetPowerStats(m_power_models[i].model_info);
+      
+      // If the timestamp is the same as the previous one (happens during startup),
+      // do not publish the data to prevent crashes.
+      if (m_power_models[i].previous_time != m_power_models[i].model_info.timestamp)
+      {
+        auto timestamp = START_TIME + std::chrono::milliseconds(
+          static_cast<unsigned>(m_power_models[i].model_info.timestamp * 1000)
+        );
+
+        double input_power, input_tmp, input_voltage;
+
+        if (start_loops > 0)
+        {
+          // The very first values sent in to each prognoser
+          // should be the init values.
+          input_power = m_initial_power;
+          input_voltage = m_initial_voltage;
+          input_tmp = m_initial_temperature;
+          start_loops--;
+        }
+        else
+        {
+          input_power = m_power_models[i].model_info.wattage;
+          input_voltage = m_power_models[i].model_info.voltage;
+          input_tmp = m_power_models[i].model_info.temperature;
+        }
+
+        // DEBUG PRINT
+        if (m_inputs_print_debug && (m_power_models[i].model_info.timestamp > 0))
+        {
+          printPrognoserInputs(input_power, input_voltage, input_tmp, i);
+        }
+
+        // Set up the input data that will be passed through the message bus
+        // to GSAP's asynchronous prognosers.
+        m_power_models[i].bus.publish(std::make_shared<DoubleMessage>(
+          MessageId::Watts, m_power_models[i].name, timestamp, input_power));
+        m_power_models[i].bus.publish(std::make_shared<DoubleMessage>(
+          MessageId::Centigrade, m_power_models[i].name, timestamp, input_tmp));
+        m_power_models[i].bus.publish(std::make_shared<DoubleMessage>(
+          MessageId::Volts, m_power_models[i].name, timestamp, input_voltage));
+
+        m_power_models[i].previous_time = m_power_models[i].model_info.timestamp;
+      }
+    }
 
-  Float64 mechanical_power_raw_msg, mechanical_power_avg_msg;
-  mechanical_power_raw_msg.data = power_watts;
-  mechanical_power_avg_msg.data = mean_mechanical_power;
-  m_mechanical_power_raw_pub.publish(mechanical_power_raw_msg);
-  m_mechanical_power_avg_pub.publish(mechanical_power_avg_msg);
+    // Print a warning message if any models computed an excessive power draw.
+    if (excessive_draw > 0)
+    {
+      ROS_WARN_STREAM("Power system computed excessive power input of "
+                      << excessive_draw << "W for one or more GSAP "
+                      << "prognosers.\nInput was capped at "
+                      << m_max_gsap_input_watts
+                      << "W for each model over the limit.");
+    }
```

Comment:
I find it odd that `excessive_draw` is not compared to `m_max_gsap_input_watts`.  In fact the latter is used only for a message.

Reply:
They are compared, just in ``PowerModelHandler`` instead of ``PowerSystemNode``.